x/txfees/keeper/hooks.go

```diff
@@ -0,0 +1,78 @@
+package keeper
+
+import (
+	// "fmt"
+
+	// "github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	epochstypes "github.com/osmosis-labs/osmosis/v7/x/epochs/types"
+	txfeestypes "github.com/osmosis-labs/osmosis/v7/x/txfees/types"
+)
+
+func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) {
+}
+
+// at the end of each epoch, swap all non-OSMO fees into OSMO and transfer to fee module account
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) {
+
+	// get module address for FooCollectorName
+	addrFoo := k.accountKeeper.GetModuleAddress(txfeestypes.FooCollectorName)
+
+	// get balances for all denoms in the module account
+	fooAccountBalances := k.bankKeeper.GetAllBalances(ctx, addrFoo)
+
+	// pulls base denom from TxFeesKeeper (should be uOSMO)
+	baseDenom, _ := k.GetBaseDenom(ctx)
+
+	// iterate through the resulting array and swap each denom into OSMO using the GAMM module
+	for _, coin := range fooAccountBalances {
+
+		if coin.Denom == baseDenom {
+			continue
+		} else {
+		
+			// TO DO: figure out how to cast this or get the pool ID for the main OSMO paired pool
+			feetoken := coin.(txfeestypes.FeeToken)
+
+			// calculate spot price to determine minimum out for swap
+			// question: is the order of input denom and output denom correct?
+			spotPrice, _ := k.spotPriceCalculator.CalculateSpotPrice(ctx, feetoken.GetPoolID(), coin.Denom, baseDenom)
+
+			// swap into OSMO
+			// question: is spotPrice really the minimum out for the swap?
+			k.spotPriceCalculator.SwapExactAmountIn(ctx, addrFoo, feetoken.GetPoolId(), coin.Denom, baseDenom, k.spotPriceCalculator)
```

Comment:
@ValarDragon Is this an appropriate place to have the swap function call?

Reply:
Yeah!

Two notes though:
* I don't know why the last argument is k.spotPriceCalculator
* We should make a second keeper in expected_keeper called GammKeeper, that we use for the Swap methods

Reply:
Sounds good, thanks! The last argument is a placeholder (see the question in the comment above it). 

How would the minimum output amount supposed to be calculated in this case?

Reply:
For now I think its fine to allow max slippage (put sdk.Coins{} as the min out)

Reply:
I believe the argument is supposed to be an integer - would it just be zero in this case?